bootloader: support execution by direct dynamic loader invocation in onedir mode

sleep-walker opened the pull request:
Onedir mode on POSIX platform performs additional exec to correctly set environment for itself. When binary is invoked as `ld-linux binary.py` and the exec is not relying on provided dynamic loader but relies instead on the one written in ELF header. We already detect this way of invocation but we didn't kept this information for the exec. This patch changes that.

bwoodsend commented:
For context: https://groups.google.com/g/pyinstaller/c/MAXaUT1i7JY

This isn't as scary as I imagined it being.

rokm commented:
Hmm, we indeed never really tested if correct dynamic loader is used, and intentionally corrupting the one stored in the ELF (e.g., changing a letter in its name with hex editor) is a good way of testing this.

That said, the proposed fix does not work for me:
``` 
$ /lib64/ld-linux-x86-64.so.2 ./dist-new/program/program
[PYI-109163:DEBUG] PyInstaller Bootloader 6.x
[PYI-109163:DEBUG] LOADER: resolved executable file /usr/lib64/ld-linux-x86-64.so.2 is ld.so dynamic loader - ignoring it! (but storing the information)
[PYI-109163:DEBUG] LOADER: resolving program path from argv[0]: ./dist-new/program/program
[PYI-109163:DEBUG] LOADER: executable file: /home/rok/tmp/pyi-exec/dist-new/program/program
[PYI-109163:DEBUG] LOADER: trying to load executable-embedded archive...
[PYI-109163:DEBUG] LOADER: attempting to open archive /home/rok/tmp/pyi-exec/dist-new/program/program
[PYI-109163:DEBUG] LOADER: cookie found at offset 0x14C7BA
[PYI-109163:DEBUG] LOADER: archive file: /home/rok/tmp/pyi-exec/dist-new/program/program
[PYI-109163:DEBUG] LOADER: application has onedir semantics...
[PYI-109163:DEBUG] LOADER: _PYI_ARCHIVE_FILE not defined...
[PYI-109163:DEBUG] LOADER: process level = 0
[PYI-109163:DEBUG] LOADER: POSIX onedir process needs to set library seach path and restart itself.
[PYI-109163:DEBUG] LOADER: setting LD_LIBRARY_PATH=/home/rok/tmp/pyi-exec/dist-new/program/_internal
./dist-new/program/program: missing program name
Try './dist-new/program/program --help' for more information.
```

That's presumably because the loader's executable should be prepended to the argv list passed to the `execv` - I'll take a closer look tomorrow.

FWIW, corrupting the dynamic loader in a onefile executable leads to the same problem as originally described, so we'll need to address that as well.

sleep-walker commented:
I verified that you're right - my testing was not performed correctly. When I tested that with simple argv printing script, it got obvious and invocation via `ld-linux app.py app.py one two three` worked as workaround.

It needs to prepend `argv` with `executable_filename` but I'll think a bit about how to make it nice (using `pyi_argv` in the end?)

rokm commented:
> It needs to prepend `argv` with `executable_filename` but I'll think a bit about how to make it nice (using `pyi_argv` in the end?)

I was thinking about this, and I would prefer to avoid using `pyi_argv` for this. Instead, I think we should go with a helper function that allocates the new argv, puts loader name in first place, then copies all the args (either from argv or pyi_argv, whichever is applicable - although in practice, pyi_argv should not be applicable on Linux), and then returns the pointer to allocated argv (or NULL). That way, all theoretical allocation failures can be handled within the helper (which can then free all allocated entries).

This helper can be then reused in both onedir and onefile codepaths. (And maybe if we made `linker_filename` available on all POSIX platforms (but used only on Linux), we can also avoid some more awkward #ifdefs in onefile codepath).

sleep-walker committed:
bootloader: support execution by direct dynamic loader invocation in onedir mode

Onedir mode on POSIX platform performs additional exec to correctly
set environment for itself. When binary is invoked as `ld-linux
binary.py` and the exec is not relying on provided dynamic loader but
relies instead on the one written in ELF header.

We already detect this way of invocation but we didn't kept this
information for the exec. This patch changes that.

sleep-walker commented:
I have prepared and tested patch which is handling arguments properly. Since `exec()` call is copying arguments and cutting stack, I haven't duplicated char* items. Thanks in advance for review.

sleep-walker commented:
I pushed tiny change to split variable declaration outside of `for` to be C99 compliant.

rokm committed:
bootloader: support execution via dynamic loader invocation, cont'd

Cosmetic fixes to previous commit.
Fix allocation of new argv to account for terminating NULL.
Add support for onefile mode.

rokm commented:
I've taken the liberty of applying cosmetic fixes and extending this PR:
- fixed the new argv allocation to account for the terminating NULL (argc + 2)
- added the relevant bits in onefile codepath

rokm committed:
bootloader: POSIX: if execvp() fails, exit immediately via exit()

If we fail to start the onefile child process via `execvp()` for
whatever reason, immediately exit the (forked) child process via
the `exit()` call. Otherwise, the child process will continue
executing the cleanup path, which is intended for the *parent*
process, and will end up interfering with the cleanup in the
actual parent process. For example, because it will remove the
temporary application directory, the parent process will fail
to do so in during its own cleanup.

rokm commented:
Thanks, @sleep-walker !
